devices/HumanDataCollector/HumanDataCollector.cpp

```diff
@@ -0,0 +1,663 @@
+    yarp::os::BufferedPort<yarp::sig::Vector> comProperAccelerationInWorldFrameDataPort;
+
+    // Yarp ports for streaming data from IHumanWrench interface of HumanWrenchProvider
+    yarp::os::BufferedPort<yarp::os::Bottle> wrenchMeasurementSourceNamesDataPort;
+    yarp::os::BufferedPort<yarp::sig::Vector> wrenchMeasurementValuesDataPort;
+
+    // Yarp ports for streaming data from IHumanWrench interfce and IHumanDynamics interface of HumanDynamicsEstimator
+    yarp::os::BufferedPort<yarp::os::Bottle> wrenchEstimateSourceNamesDataPort;
+    yarp::os::BufferedPort<yarp::sig::Vector> wrenchEstimateValuesDataPort;
+
+    yarp::os::BufferedPort<yarp::os::Bottle> dynamicsJointNamesDataPort;
+    yarp::os::BufferedPort<yarp::sig::Vector> jointTorquesDataPort;
+};
+
+HumanDataCollector::HumanDataCollector()
+    : PeriodicThread(DefaultPeriod)
+    , pImpl{new impl()}
+{}
+
+HumanDataCollector::~HumanDataCollector() {}
+
+bool HumanDataCollector::open(yarp::os::Searchable &config) {
+
+    // ===============================
+    // CHECK THE CONFIGURATION OPTIONS
+    // ===============================
+
+    if (!(config.check("period") && config.find("period").isFloat64())) {
+        yInfo() << LogPrefix << "Using default period";
+    }
+
+    if (!(config.check("portPrefix") && config.find("portPrefix").isString())) {
+        yInfo() << LogPrefix << "Using default portPrefix HDE ";
+    }
+
+    // ===============================
+    // PARSE THE CONFIGURATION OPTIONS
+    // ===============================
+
+    double period = config.check("period", yarp::os::Value(0.01)).asFloat64();
+    pImpl->portPrefix = config.check("portPrefix", yarp::os::Value("HDE")).asString();
+
+    //TODO: Define rpc to reset the data dump or restarting the visualization??
+
+    yInfo() << LogPrefix << "*** ===========================";
+    yInfo() << LogPrefix << "*** Period                    :" << period;
+    yInfo() << LogPrefix << "*** portPrefix                :" << pImpl->portPrefix;
+    yInfo() << LogPrefix << "*** ===========================";
+
+    // Set period
+    setPeriod(pImpl->period);
+
+    return true;
+}
+
+bool HumanDataCollector::close() {
+    return true;
+}
+
+void HumanDataCollector::threadRelease() {}
+
+void HumanDataCollector::run()
+{
+    if (pImpl->firstData) {
+
+        // =====================================================================
+        // Open yarp ports base on the attached devices for streaming human data
+        // =====================================================================
+
+        if (pImpl->isStateProviderDevice) {
+
+            // Open base pose data port
+            const std::string basePosePortName = "/" + pImpl->portPrefix + "/" + DeviceName + "/basePose:o";
+            if (!pImpl->basePoseDataPort.open(basePosePortName)) {
+                yError() << LogPrefix << "Failed to open port " << basePosePortName;
+                askToStop();
```

Comment:
I am not sure, if you want to stop, probably returning and avoid continuing the execution would be the correct flow?

```cpp
if (...) {
    // ...
    askToStop();
    return;
}
```

Reply:
@diegoferigo `askToStop()` is the suggested function to be called from `run()` https://www.yarp.it/classyarp_1_1os_1_1PeriodicThread.html#aa92156efefdf172da41077d5962747a2

Reply:
Yes but it does not return from the function. I'm not saying to remove it, I'm saying to add the `return` :)